### PR TITLE
[flink] Support to set the table option by dynamic global job configu…

### LINKED ISSUE #2037: [Feature] Support to set the table option by dynamic global job configuration
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

When using Flink to query the Paimon table, supporting overwriting the catalog table option by `SET` syntax will be more convenient. The dynamic options can only take effect in the current session.

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Aitozi**
…ration

### Purpose

Linked issue: close #2037

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java [hunk 1]
```diff
@@ -263,4 +286,57 @@ static boolean schemaEquals(RowType rowType1, RowType rowType2) {
...
+     * The dynamic option's format is:
+     *
+     * <p>{@link
+     * FlinkConnectorOptions#TABLE_DYNAMIC_OPTION_PREFIX}.${catalog}.${database}.${tableName}.key =
+     * value. These job level configs will be extracted and injected into the target table option.
+     *
+     * @param context The table factory context.
+     * @return The dynamic options of this target table.
+     */
+    static Map<String, String> getDynamicTableConfigOptions(DynamicTableFactory.Context context) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:300]: Can you also document this in flink page?

Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:300]: Updated

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java [hunk 2]
```diff
@@ -263,4 +286,57 @@ static boolean schemaEquals(RowType rowType1, RowType rowType2) {
...
+                        FlinkConnectorOptions.TABLE_DYNAMIC_OPTION_PREFIX,
+                        context.getObjectIdentifier().getCatalogName(),
+                        context.getObjectIdentifier().getDatabaseName(),
+                        context.getObjectIdentifier().getObjectName());
+        Pattern pattern = Pattern.compile(template);
+
+        conf.keySet()
+                .forEach(
+                        (key) -> {
+                            Matcher matcher = pattern.matcher(key);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:328]: `pattern.matcher` is very slow, can we just check starts with `paimon.` first?

File: docs/content/engines/flink.md [hunk 3]
```diff
@@ -301,3 +301,19 @@ Users can set memory weight in SQL for Flink Managed Memory, then Flink sink ope
...
+
+For example:
+
+```sql
+-- set scan.timestamp-millis=1697018249000 for the table mycatalog.default.T
+SET 'paimon.mycatalog.default.T.scan.timestamp-millis' = '1697018249000';
+SELECT * FROM T;
+
+-- set scan.timestamp-millis=1697018249000 for the table default.T in any catalog
+SET 'paimon.*.default.T.scan.timestamp-millis' = '1697018249000';
```
Comment (JingsongLi) [inline docs/content/engines/flink.md:317]: Can we just support paimon.{key} for all tables?

Comment (Aitozi) [reply docs/content/engines/flink.md:317]: for all table may cause unexpected behavior, if we use two table but only want to set option to one of them

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java (hunk 2 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:328]: It's only happened in client planner phase and the pattern is not complex

Comment (JingsongLi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:328]: You can do a little benchmark for these two methods, and I think there may be hundreds options need to be checked.

File: docs/content/engines/flink.md (hunk 3 above)
Comment (JingsongLi) [reply docs/content/engines/flink.md:317]: Yes, there is unexpected behavior, but there is also user case to define options for all tables.

Comment (JingsongLi) [reply docs/content/engines/flink.md:317]: Maybe not `scan.timestamp-millis`, but some options about execution.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java (hunk 2 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:328]: ok, will do a test

File: docs/content/engines/flink.md (hunk 3 above)
Comment (Aitozi) [reply docs/content/engines/flink.md:317]: > Yes, there is unexpected behavior, but there is also user case to define options for all tables.

if user want to define for all tables they can set `paimon.*.*.*.config_key=value`. This means all tables will take this config_key. if user want to set options under a database, they can set `paimon.*.db.*.config_key=value`. They just have to keep in mind that these three part controls the catalog, database, and table scope.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java (hunk 2 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java:328]: OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Mac OS X 10.16
Apple M1 Pro
test-pattern:                                                                                        Best/Avg Time(ms)    Row Rate(M/s)      Per Row(ns)   Relative
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
OPERATORTEST_test-pattern_pattern                                                                       12530 / 12779              0.0          41767.0       1.0X
OPERATORTEST_test-pattern_startswith                                                                       500 /  538              0.6           1666.2      25.1X

Per row will run for 100 key. so the max cost for 100 key match is 41767 ns. The startsWith is much faster, but I think both of the cost is acceptable, and the startWith check is not flexible when we only want to set  for the specific table

**Comment by Aitozi** [comment]
Updated, Please take a look again @JingsongLi

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**